### PR TITLE
Migrate messages to ICU number skeletons

### PR DESCRIPTION
**Pull request by rszwajko**
Before, custom functions were used to format parameters passed to
translated messages. With this patch the formatting in such cases is
moved into the translation layer.

Intl.NumberFormat properties can be mapped to following concise number
skeleton syntax:
1. minimumIntegerDigits: 1 -> 0
2. minimumFractionDigits: 0, maximumFractionDigits: 0 -> .
3. minimumFractionDigits: 1, maximumFractionDigits: 1 -> .0
4. useGrouping: false -> ,_
5. style: 'percent' -> %

Bug-Url: https://bugzilla.redhat.com/2024202

### CODE REVIEW

**Comment by rszwajko** [comment]
## Dashboard with number skeleton based formatting (Spanish locale)
image

**Comment by rszwajko** [comment]
@sjd78 @sgratch 
This PR requires newer versions of intl libraries. For simplicity the same versions were used as in VM Portal so that we could re-use the same ovirt-engine-nodejs-modules version. Please let me know if such upgrade is OK at this point.

File: package.json [hunk 1]
```diff
@@ -37,7 +37,7 @@
     "css-element-queries": "^0.4.0",
     "d3": "^3.5.16",
     "intl": "^1.2.4",
-    "intl-messageformat": "^7.2.1",
+    "intl-messageformat": "^9.9.4",
```
Comment (sjd78) [inline package.json:40]: Technically `^9.9.4` is not the same as `9.9.4` used in ovirt-web-ui.  When you added it, yarn picked up `9.11.4`.  That's why the offline build fails.

File: TRANSLATION_GUIDE.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ ### Number skeleton
+
+Common number formatting options (in concise syntax) and their mapping to Intl.NumberFormat properties (used for custom formatting functions):
+1. `minimumIntegerDigits: 1` -> `0` ([integer-width](https://unicode-org.github.io/icu/userguide/format_parse/numbers/skeletons.html#integer-width))
+2. `minimumFractionDigits: 0`, `maximumFractionDigits: 0` -> `.` (equivalent to [precision-integer](https://unicode-org.github.io/icu/userguide/format_parse/numbers/skeletons.html#precision) )
+3. `minimumFractionDigits: 1`, `maximumFractionDigits: 1` -> `.0` ([fraction-precision](https://unicode-org.github.io/icu/userguide/format_parse/numbers/skeletons.html#fraction-precision))
+4. `useGrouping: false` -> `,_` ([group-off](https://unicode-org.github.io/icu/userguide/format_parse/numbers/skeletons.html#grouping))
+5. `style: 'percent'` -> `%` ([percent](https://unicode-org.github.io/icu/userguide/format_parse/numbers/skeletons.html#unit) unit)
+
+Example:
```
Comment (sjd78) [inline TRANSLATION_GUIDE.md:19]: Add an anticipated result in addition to what the format will look like.  Just to connect the doc, example and output together...for example:

"Given the following message:
```js
{
  demoFormat: {
    id: 'demoFormat',
    defaultMessage: 'This shows {value, number, ......} percent as the  formatted output',
  }
}
```

when we call `msg.demoFormat({ value: 0.125 })` the output will be:
```
  This shows 13% as the formatted output.
```
"

**Review by sjd78 — CHANGES_REQUESTED** [review]
The approach looks pretty good.

A few thoughts:

1. Add an example or two to the translation guide, or maybe the common ones you used.  That will help make it easy to understand the ICU formatting.

2. Looks like you have a few common cases for formatting used in the messages.  It is probably prudent to add those as unit tests somewhere to make sure they keep rendering as anticipated.  Fun to note...the JS rounding code is slightly different in different places so the auto rounding results can be different between nodejs and browser runtimes.  Or maybe not anymore?  I haven't had to look at it for awhile.

3. Any thought about just doing a custom format for the common formatting used throughout as a shorthand?

**Comment by rszwajko** [comment]
@sjd78 
> Any thought about just doing a custom format for the common formatting used throughout as a shorthand?
>

This was my first approach (as you suggested). However custom formats create some problems:
1. they are meaningful only locally, inside our project - translators won't know their meaning.
2. it's hard to create shorter meaningful names then concise ICU syntax  which uses (almost) single char per feature i.e. function name `formatNumber1D()` contains much less information than the corresponding format string  `:: .0 0 ,_`.  
3. custom formats make harder to use 3rd party intl tooling i.e. I've encountered problems with our parse tests which use vanilla `intl-messageformat-parser` and were not aware of custom formats. 
4. we can run into name conflicts i.e. the already used custom format `0` uses the same name as ICU integer-width stem and custom format `0.0` overlaps with ICU fraction-precision stem.

**Comment by sjd78** [comment]
> @sjd78
> 
> > Any thought about just doing a custom format for the common formatting used throughout as a shorthand?
> 
> This was my first approach (as you suggested). However custom formats create some problems:
> 
>     1. they are meaningful only locally, inside our project - translators won't know their meaning.
> 
>     2. it's hard to create shorter meaningful names then concise ICU syntax  which uses (almost) single char per feature i.e. function name `formatNumber1D()` contains much less information than the corresponding format string  `:: .0 0 ,_`.
> 
>     3. custom formats make harder to use 3rd party intl tooling i.e. I've encountered problems with our parse tests which use vanilla `intl-messageformat-parser` and were not aware of custom formats.
> 
>     4. we can run into name conflicts i.e. the already used custom format `0` uses the same name as ICU integer-width stem and custom format `0.0` overlaps with ICU fraction-precision stem.

Cool.  Makes total sense to go the direction you did.

File: TRANSLATION_GUIDE.md (hunk 2 above)
Comment (sgratch) [reply TRANSLATION_GUIDE.md:19]: You didn't add an output for the example, as Scott suggested. I think it worth adding....

**Review by sgratch — COMMENTED** [review]
LGTM, except the screenshot attached above which shows `dashboardUtilizationCardOverCommit` numbers (`overcommit`, `allocated`) as > 100%. 
This is wrong. Is this the same for English locale? Is it fake data or a real one?

> image

**Comment by rszwajko** [comment]
@sgratch 
> LGTM, except the screenshot attached above which shows `dashboardUtilizationCardOverCommit` numbers (`overcommit`, `allocated`) as > 100%.
> This is wrong. Is this the same for English locale? Is it fake data or a real one?
>

It's fake data.

File: TRANSLATION_GUIDE.md (hunk 2 above)
Comment (sgratch) [reply TRANSLATION_GUIDE.md:19]: @rszwajko are you planning to add an example, not something fancy, just show that the msg example mentioned below for  `dashboardUtilizationCardAvailableOfPercent` can be translated to:....if `percent` is set to....

**Comment by sgratch** [comment]
> @sjd78 @sgratch Not sure how well we should describe/test the ICU format - the official documentation is quite good. IMHO it make no sense to duplicate it (at least we don't do it for other libraries/frameworks like React, Redux, Sagas etc).

I thought just to show the result of your example in the TRANSLATE_GUIDE, not something fancy. Just showing that the msg example mentioned for `dashboardUtilizationCardAvailableOfPercent` can be translated to:....if percent value is set to....

File: TRANSLATION_GUIDE.md (hunk 2 above)
Comment (rszwajko) [reply TRANSLATION_GUIDE.md]: Sure. Added the output.

**Comment by rszwajko** [comment]
@sjd78 
> Looks like you have a few common cases for formatting used in the messages. It is probably prudent to add those as unit tests somewhere to make sure they keep rendering as anticipated
>

Can we do it in follow up PR? Unit testing intl required some extra time.

**Review by sgratch — APPROVED** [review]
LGTM

**Review by sjd78 — APPROVED** [review]
LGTM

**Pull request merged**